flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/PlannerCallProcedureOperation.java

```diff
@@ -0,0 +1,379 @@
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.conversion.DataStructureConverter;
+import org.apache.flink.table.data.conversion.DataStructureConverters;
+import org.apache.flink.table.operations.CallProcedureOperation;
+import org.apache.flink.table.operations.Operation;
+import org.apache.flink.table.operations.OperationUtils;
+import org.apache.flink.table.planner.functions.casting.RowDataToStringConverterImpl;
+import org.apache.flink.table.procedure.DefaultProcedureContext;
+import org.apache.flink.table.procedure.ProcedureContext;
+import org.apache.flink.table.procedures.Procedure;
+import org.apache.flink.table.procedures.ProcedureDefinition;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.extraction.ExtractionUtils;
+import org.apache.flink.table.types.logical.utils.LogicalTypeChecks;
+import org.apache.flink.table.types.utils.DataTypeUtils;
+import org.apache.flink.table.utils.print.RowDataToStringConverter;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.CloseableIterator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.time.ZoneId;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.types.extraction.ExtractionUtils.isAssignable;
+
+/** Wrapper for valid call procedure operation generated by Planner. */
+public class PlannerCallProcedureOperation implements CallProcedureOperation {
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(PlannerCallProcedureOperation.class);
+
+    private final ObjectIdentifier procedureIdentifier;
+    private final Procedure procedure;
+
+    /** The internal represent for input arguments. */
+    private final Object[] internalInputArguments;
+
+    private final DataType[] inputTypes;
+    private final DataType outputType;
+
+    public PlannerCallProcedureOperation(
+            ObjectIdentifier procedureIdentifier,
+            Procedure procedure,
+            Object[] internalInputArguments,
+            DataType[] inputTypes,
+            DataType outputType) {
+        this.procedureIdentifier = procedureIdentifier;
+        this.procedure = procedure;
+        this.internalInputArguments = internalInputArguments;
+        this.inputTypes = inputTypes;
+        this.outputType = outputType;
+    }
+
+    @Override
+    public TableResultInternal execute(Context ctx) {
+        TableConfig tableConfig = ctx.getTableConfig();
+        ClassLoader userClassLoader = ctx.getResourceManager().getUserClassLoader();
+
+        // get the class for the args
+        Class<?>[] argumentClz = new Class[1 + inputTypes.length];
```

Comment:
I might have a little concerns here. Users may have a certain level of understanding cost because the parameter list declared in the producer is inconsistent with the actual parameter list used, as we default to using "context" as the first parameter.  

WDYT ?

Reply:
For procedure devs, it need to understand it. 
But for users, actually, I think they don't need to understand it.
The procedure devs develops the procedure with a method like `call(Context c, arg1 int, arg2 int)`
and then documents for users how to use it like using the statement `call(1, 2)`. The user actaully don't need to care about the first wired `context` parameter. Flink framework will just adjust the arguments and involve the method `call(Context c, arg1 int, arg2 int)`.  It's transparent to users.